[cosmos] add gas limit estimation for transfer, rewards, delegations, undelegations and redelegations

barjonr opened the pull request.

Commit:
[cosmos] add gas limit estimation for messages of type MSGSEND

Commit:
[cosmos] gas limit estimation

For transfer, delegation, undelegation,
redelegation and getting rewards.

gagbo commented on core/src/wallet/cosmos/api_impl/CosmosLikeTransactionApi.cpp:
```diff
@@ -340,5 +340,99 @@ namespace ledger {
+            auto baseReq = rapidjson::Value(rapidjson::kObjectType);
+            for (const auto &message : _txData.messages) {
+                const auto tmpMessage = std::make_shared<CosmosLikeMessage>(message);
+                if (!(tmpMessage->getMessageType() == api::CosmosLikeMsgType::MSGSEND)) continue;
```
This means currently the gas estimation will only serialize transfers to send the simulate request ?

barjonr replied:
I forgot to remove this function for CosmosLikeTransactonApi. It has been moved to the Gaia explorer with the functions to serialize other messages (delegation, rewards, undelegation, redelegation). I will remove it.

gagbo commented on core/src/wallet/cosmos/api_impl/CosmosLikeTransactionApi.cpp:
```diff
@@ -340,5 +340,99 @@ namespace ledger {
+                amount.PushBack(amountObj, allocator);
+                document.AddMember(cosmos::constants::kAmount, amount, allocator);
+
+                break;
```
Once this works, we'll have to refactor it, probably in static `CosmosLikeMessage` functions, it gets hard to read fast otherwise :)

barjonr replied:
Same than the comment above, I will remove this

gagbo commented on core/src/wallet/cosmos/api_impl/CosmosLikeTransactionApi.cpp:
```diff
@@ -340,5 +340,99 @@ namespace ledger {
+                    tmp.SetString(denom.c_str(), static_cast<rapidjson::SizeType>(denom.length()), allocator);
+                    amountObject.AddMember(cosmos::constants::kDenom, tmp, allocator);
+                    return amountObject;
+                };
```
Looks like the only capture is the allocator. We might be able to extract that into a real function later

barjonr replied:
I have done that in the explorer, this code will be removed, I just forgot to do it

gagbo commented on core/src/wallet/cosmos/explorers/GaiaCosmosLikeBlockchainExplorer.cpp:
```diff
@@ -427,5 +433,356 @@ namespace ledger {
 
+        namespace {
+
+        // TODO: probably move most of these in TransactionApi
```
If we move them, I'm for creating a new folder altogether just for json serialization/deserialization. The code is in too many places now in my opinion :)

gagbo replied:
Maybe a `GaiaJsonLayer` or something ?

barjonr replied:
I agree, we should gather all these things in the same place, it will probably be easier to use

gagbo replied:
We'll do that after getting send/receive/gas working I suppose

gagbo commented on core/src/wallet/cosmos/explorers/GaiaCosmosLikeBlockchainExplorer.cpp:
```diff
@@ -427,5 +433,356 @@ namespace ledger {
+
+            auto buffer = rapidjson::StringBuffer();
+            auto writer = rapidjson::Writer<rapidjson::StringBuffer>(buffer);
+            // TODO: sortJson is defined static in CosmosLikeTransactionApi.cpp (either make it visible or move this to CosmosLikeTransactionApi)
+            // sortJson(document);
```
Since the device does not need to sign the payload here, sorting the json should not matter here.

But having sortJson as a hidden function is not so good I agree. Maybe we'll move it up to `utils` or something.

gagbo commented on core/src/wallet/cosmos/explorers/GaiaCosmosLikeBlockchainExplorer.cpp:
```diff
@@ -427,5 +433,356 @@ namespace ledger {
+            return async::sequence(getContext(), estimations)
+                .flatMapPtr<BigInt>(getContext(), [](const auto& estimations) {
+                        const auto result = std::accumulate(std::begin(estimations),
+                                std::end(estimations), BigInt::ZERO);
```
Can you just add a comment to say that signature verification cost is accumulated 1 time per message in the transaction, but that we don't care because it's usually a lot smaller than the rest of the gas cost ?

I'm afraid we will forget otherwise (or maybe I'm wrong and I forgot what we decided about fixed gas costs)

barjonr replied:
I can do that, or I can remove the cost of the signature (with the constant value in the cosmos repo) for every extra message. Just tell me what you prefer.

gagbo replied:
Let's ignore the cost for now, common transactions use amounts in the `x e4` Gas, the signature cost can be ignored I think

gagbo reviewed (changes requested):
Looks great and well organized :) I have only one comment about an early `continue` in serializeTransaction, but most other comments are markers for the future refactoring work we will probably have to do

gagbo commented on core/test/integration/CMakeLists.txt:
```diff
@@ -212,3 +212,4 @@ add_test (NAME ledger-core-integration-CosmosLikeWalletSynchronization.GetSendWi
 add_test (NAME ledger-core-integration-CosmosLikeWalletSynchronization.GetDelegateWithExplorer COMMAND ledger-core-integration-tests --gtest_filter=CosmosLikeWalletSynchronization.GetDelegateWithExplorer)
 add_test (NAME ledger-core-integration-CosmosLikeWalletSynchronization.GetCurrentBlockWithExplorer COMMAND ledger-core-integration-tests --gtest_filter=CosmosLikeWalletSynchronization.GetCurrentBlockWithExplorer)
 add_test (NAME ledger-core-integration-CosmosLikeWalletSynchronization.AllTransactionsSynchronization COMMAND ledger-core-integration-tests --gtest_filter=CosmosLikeWalletSynchronization.AllTransactionsSynchronization)
+add_test (NAME ledger-core-integration-CosmosLikeWalletSynchronization.GasLimitEstimation COMMAND ledger-core-integration-tests --gtest_filter=CosmosLikeWalletSynchronization.GasLimitEstimation)
```
The tests added in cosmos_synchronization.cpp aren't all in this list.

(Last time I checked, we used that instead of automatic discovery because of some esoteric Qt::HttpClient related bug)

barjonr replied:
I also forgot to add them, I will do it right now

Commit:
[cosmos] small changes after the code review

gagbo reviewed (approved):
Looks good thanks :)

ghost commented:
Do your new tests pass?
I tried to run them to figure out if `accountNumber` and `accountSequence` are actually needed or not (see PR https://github.com/gagbo/lib-ledger-core/pull/27), but I seem to get `Bad Request` every time?

haammar-ledger commented on core/src/wallet/cosmos/explorers/GaiaCosmosLikeBlockchainExplorer.cpp:
```diff
@@ -427,5 +433,363 @@ namespace ledger {
+        };
+
+        template <typename... JsonObj>
+        std::string makeJsonFrom(rapidjson::Document& document,
```
There's an issue on my side with this method, the resulting output looks like `"{\"base_req\":null,\"delegator_address\":{\"amount\":\"1000000\",\"denom\":\"uatom\"},\"validator_address\":null,\"amount\":null}"` for example.

I can't understand why and if it's something on my side. We can always merge and test on the main branch...

ghost commented:
(Still need to solve conflicts before merging I guess)

Commit:
[cosmos] fix serialization when compiling with g++

Changed the signatures of utility functions `makeXXX`
from `rapidjson::Value makeXXX(...)` to
`void makeXXX(..., rapidjson::Value&, ...)`

gagbo reviewed (approved):
LGTM :+1:
